src/particle.cpp

```diff
@@ -682,10 +683,29 @@ Particle::write_restart() const
     write_dataset(file_id, "type", static_cast<int>(type_));
 
     int64_t i = current_work_;
-    write_dataset(file_id, "weight", simulation::source_bank[i-1].wgt);
-    write_dataset(file_id, "energy", simulation::source_bank[i-1].E);
-    write_dataset(file_id, "xyz", simulation::source_bank[i-1].r);
-    write_dataset(file_id, "uvw", simulation::source_bank[i-1].u);
+    if (settings::run_mode == RunMode::EIGENVALUE) {
+      //take source data from primary bank for eigenvalue simulation
+      write_dataset(file_id, "weight", simulation::source_bank[i-1].wgt);
+      write_dataset(file_id, "energy", simulation::source_bank[i-1].E);
+      write_dataset(file_id, "xyz", simulation::source_bank[i-1].r);
+      write_dataset(file_id, "uvw", simulation::source_bank[i-1].u);
+    } else if (settings::run_mode == RunMode::FIXED_SOURCE) {
+      // re-sample using rng random number seed used to generate source particle
+      int64_t id = (simulation::total_gen + overall_generation() - 1)*settings::n_particles +
+        simulation::work_index[mpi::rank] + i;
+      uint64_t seed = init_seed(id, STREAM_SOURCE);
+      // re-sample source site
+      Particle::Bank site;
+      if (!settings::path_source_library.empty()) {
+        site = sample_custom_source_library(&seed);
+      } else {
+        site = sample_external_source(&seed);
+      }
```

Comment:
This block of code appears exactly the same here and in simulation.cpp. What I would do is just change `sample_external_source` to have a block at the beginning like:
```C++
if (!settings::path_source_libary.empty()) {
  return sample_custom_source_library(seed)
}
```
Then the code here would just become
```suggestion
      auto site = sample_external_source(&seed);
```

Reply:
Done